Feature #29 상품 등록

KJW921222 opened the pull request:
선정한 패키지 구조로 패키지를 생성하고, 상품을 등록하는 기능을 구현한다.

f-lab-robert committed:
Merge pull request #32 from f-lab-edu/feature-#31

Issues-#31 프로젝트 세팅

KJW921222 committed:
프로젝트 세팅

########
1. 패키지 구조 선정, 패키지 생성

########
issue#33

KJW921222 committed:
프로젝트 세팅

########
1. 패키지 구조 선정, 패키지 생성

########
issue#33

KJW921222 committed:
프로젝트 세팅

########
1. 패키지 구조 선정, 패키지 생성

########
issue#33

KJW921222 committed:
프로젝트 세팅

########
1. 패키지 구조 선정, 패키지 생성

########
issue#33

KJW921222 committed:
프로젝트 세팅

########
1. 패키지 구조 선정, 패키지 생성

########
issue#33

KJW921222 committed:
상품 등록

########
1.

########

KJW921222 committed:
상품 등록

########
1.

########

KJW921222 committed:
상품 등록

########
1.

########

KJW921222 commented:
@f-lab-robert 
성권님과 리뷰 후 다음과 같은 개선사항을 확인하였습니다.

1.  데이터 모델 객체를 표현하는 패키지 및 클래스 추가

KJW921222 committed:
상품 등록

########
1.

########

KJW921222 committed:
상품 등록

########
1.

########

KJW921222 committed:
상품 등록

########
1.

########

KJW921222 committed:
상품 등록

########
1.

########

JinXuanT commented:
니엄마

JinXuanT commented:
니아빠

KJW921222 committed:
상품 등록

########

########

KJW921222 committed:
상품 등록

########

########

KJW921222 committed:
상품 등록

########

########

f-lab-robert commented on src/test/java/com/flabedu/small/small/service/ItemServiceTest.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    @Test
+    @DisplayName("상품 등록")
```
상품 등록 시 필요한 메소드가 모두 호출 된다. 

위의 명칭으로 바뀌는게 더 적합해보입니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/ItemControllerTest.java:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    @Test
+    @DisplayName("/items POST 상품 등록 실패")
```
상품 등록 실패로 이름을 바꾸셔도 될 것 같습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/dto/ItemDTO.java:
```diff
@@ -0,0 +1,46 @@
+
+@Getter
+@Builder
+public class ItemDTO {
```
ItemDTO 는 사용자에게서 요청을 받을 때 사용하는 DTO 아닌가요 ?

그러면 itemId 가 없는게 적합해보입니다. 참고로 requestDTO 는 사용자의 요청을 받을 때 필요한 컬럼들만 있어야 합니다. 개발자가 하나 크게 클래스를 만들어서 모두 필요한 곳에 ( 가령 query 날릴 때, 계층 이동할 때, 사용자 요청 받을 때 ) 이용하기 위해서 모든 칼람을 다 한곳에 모아 놓고 사용하면 안됩니다.

실무에서는 API 명세서를 자동으로 generate 를 해주는 swagger 를 자주 이용하는데요. swagger 는 주원님이 많든 API 의 input 에 itemID 가 필수라고 사용자에게 가이드해줄 것 입니다. 때문에 수정해주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/controller/ItemController.java:
```diff
@@ -0,0 +1,34 @@
+        return ResponseEntity.status(HttpStatus.CREATED).build();
+    }
+
+
```
불필요한 띄어씌기 등은 다 제거해주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/web/controller/ItemController.java:
```diff
@@ -0,0 +1,34 @@
+
+
+@RequestMapping("/items")
+@Slf4j
```
이 어노테이션은 쓰신 이유가 무엇일까요 ? 어떤 역할을 해주는건가요

KJW921222 replied:
Slf4j는 Logback, Log4j2 등의 로깅 프레임워크를 일관된 방식으로 사용할 수 있도록 해주는 PSA입니다.
@Slf4j 어노테이션은 롬복에서 제공해주고, 로거 객체를 생성하는 코드의 반복을 줄여주는 역할을 합니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/service/ItemService.java:
```diff
@@ -0,0 +1,41 @@
+                .build();
+
+        itemRepository.addItem(item);
+
```
불필요한 공백 아닌가요

f-lab-robert commented on src/main/java/com/flabedu/small/small/repository/ItemRepository.java:
```diff
@@ -0,0 +1,19 @@
+@Mapper
+@Repository
+public interface ItemRepository {
+    Long addItem(Item item);
```
Item 또한 DTO 처럼 더 세분화되서 관리가 되어야 하지 않나요 ?

KJW921222 replied:
item_image, item_category, item_detail 도 model 클래스를 작성하라는 말씀이 맞을까요??

f-lab-robert commented:
Test Case 가 더 많이 추가되어야할 것 같습니다. 다음 상황들에 대해 Test Case 를 작성해보시겠어요 ?

1. API 사용자가 DTO 에 필요한 컬럼등을 제대로 기재하지 않은 채로 API 로 요청하는 경우 ( 이 경우 주원님이 개발한 API 가 어떤 형태의 어떤 에러를 응답으로 보내는지 체크해봐야겠죠 ? )
2. mapper 에서 에러가 났을 경우, 어떤 상황이 발생하고 그게 주원님이 기대한 상황인지

KJW921222 committed:
상품 등록

########

########

KJW921222 committed:
상품 등록

########

########

KJW921222 committed:
상품 등록

########

########

KJW921222 commented:
ItemDetailDTO 존재 필요성

- ItemDetail은 Item 테이블에 종속된 테이블이고 not null 필드만 가지고 있어,  DTO를 생성하는 것보다 model만 사용하는 게 적합하다고 생각함. 

Enum 클래스 위치

1.  상수(혹은 enum) 패키지를 생성하고, 그 하위에 위치시킨다.
2. model 패키지의 하위에 위치시킨다.
3. model 패키지 하위에 상수(혹은 enum) 패키지를 생성하고, 그 하위에 위치시킨다.
-> 우선 2번을 선택했습니다

에러코드 정의

-  예제를 구글링해보고, 다른 팀들의 프로젝트를 많이 참고해봤는데요. 예상 가능한 예외들을 미리 enum으로 정의해놓고 CustomException을 던지는 케이스가 많았습니다. 성권님과 상의 후 적용해보아도 될까요?

f-lab-robert commented:
@KJW921222 

> ItemDetail은 Item 테이블에 종속된 테이블이고 not null 필드만 가지고 있어, DTO를 생성하는 것보다 model만 사용하는 게 적합하다고 생각함.

- 계층이 다릅니다. Service 로직이 적다면 ServiceClass 를 만들지 않고 Controller 에서 모든 것을 처리하는게 맞을까요 ? 계층간의 구분을 지으려는 이유가 최초에 어떤게 있으셨을까요 ?

> Enum 클래스 위치

- 마찬가지입니다. 두 분이 선택하신게 계층형 아니셨나요 ? Enum 이라는 기능의 관점이 아닌 계층의 관점에서 생각해보시겠어요 ?

> 에러코드 정의

- 시작점을 다시 잡아주세요. CustomError 가 필요한 이유. 어떻게 다른 프로젝트나 실무에서는 필요로 하고 적용하는지 부터 시작해주세요.

KJW921222 committed:
## 작업사항
1. 불필요한 공백 제거
2. 테스트 코드 @DisplayName 수정
3. ItemDTO의 itemId 제거
4. dto - itemDetailDTO 클래스 추가, itemDetail로 변환하는 메소드 작성
5. model - ItemImage , ItemCategory 클래스 추가

KJW921222 committed:
## 작업사항
1. 불필요한 공백 제거
2. 테스트 코드 @DisplayName 수정
3. ItemDTO의 itemId 제거
4. dto - itemDetailDTO 클래스 추가
   * itemDetail로 변환하는 메소드 작성
5. model - ItemImage , ItemCategory 클래스 추가
6. ErrorCode Enum 클래스 추가
   * CutsomException 필드 수정
   * GlobalExceptionHandler 수정
7. model - enums 패키지 생성
   * GenderEnum, SizeEnum 이동
8. ItemControllerTest 수정
   * 부족한 파라미터 개수로 인한 상품 등록 실패 테스트 케이스 작성

 ## 기타사항
 패키지 구성 방법을 계층형 구조로 선정했기 때문에, enum을 사용하는 계층에 enum 클래스를 넣어두는 것이 가장 적합하다고 생각하여 model 하위에 enum 클래스를 위치시키도록 함. 계층형 구조이므로 model에 너무 많은 클래스가 작성될 수 있어, model - enums 패키지를 생성하여 그 하위에 enum 클래스를 위치시킴.

KJW921222 committed:
## 작업사항
1. 불필요한 공백 제거
2. 테스트 코드 @DisplayName 수정
3. ItemDTO의 itemId 제거
4. dto - itemDetailDTO 클래스 추가
   * itemDetail로 변환하는 메소드 작성
5. model - ItemImage , ItemCategory 클래스 추가
6. ErrorCode Enum 클래스 추가
   * CutsomException 필드 수정
   * GlobalExceptionHandler 수정
7. model - enums 패키지 생성
   * GenderEnum, SizeEnum 이동
8. ItemControllerTest 수정
   * 부족한 파라미터 개수로 인한 상품 등록 실패 테스트 케이스 작성

 ## 기타사항
 패키지 구성 방법을 계층형 구조로 선정했기 때문에, enum을 사용하는 계층에 enum 클래스를 넣어두는 것이 가장 적합하다고 생각하여 model 하위에 enum 클래스를 위치시키도록 함. 계층형 구조이므로 model에 너무 많은 클래스가 작성될 수 있어, model - enums 패키지를 생성하여 그 하위에 enum 클래스를 위치시킴.

KJW921222 committed:
2022.06.01

KJW921222 commented:
@f-lab-robert 
enum에 포함되지 않거나 타입이 맞지 않는 파라미터를 강제로 보냈을 때, HttpMessageNotReadableException 이 발생하여 exceptionhandler로 잡았습니다. 
그런데 테스트 코드를 작성해서 검증해보려하니 자료형이 안맞아서 mock 객체가 생성이 안되는데, 이런 경우에 테스트 코드는 어떻게 작성해야 하나요?? 이런 케이스는 체크하는 방식이 따로 있는지 궁금합니다

f-lab-robert commented:
> enum에 포함되지 않거나 타입이 맞지 않는 파라미터를 강제로 보냈을 때, HttpMessageNotReadableException 이 발생하여 exceptionhandler로 잡았습니다.
> 그런데 테스트 코드를 작성해서 검증해보려하니 자료형이 안맞아서 mock 객체가 생성이 안되는데, 이런 경우에 테스트 코드는 어떻게 작성해야 하나요?? 이런 케이스는 체크하는 방식이 따로 있는지 궁금합니다

: 매번 말씀드렸지만 PR 입니다. 구체적인 코드 라인으로 예시를 들어주시거나, 예시를 직접 작성해서 질문해주세요. 주원님의 머릿 속에 있는 것들을 문장 몇 줄로 PR 참여자들이 모두 이해하기를 바라는건 바람직하지 못합니다. 비슷한 코멘트를 많이 드렸던 것 같습니다. 질문에 시간을 많이 쏟아주세요.

f-lab-robert commented on src/main/java/com/flabedu/small/small/exception/CustomErrorResponse.java:
```diff
@@ -0,0 +1,14 @@
+public class CustomErrorResponse {
+    private int code;
+    private String message;
+    private List<CustomFieldError> error;
```
> private List<CustomFieldError> error;


이 부분이 왜 필요한걸까요. code 와 message 로 충분하지 않나요 ?

f-lab-robert commented on src/main/java/com/flabedu/small/small/model/Item.java:
```diff
@@ -0,0 +1,26 @@
+    private List<ItemDetail> itemDetails;
+    private String registUserId;
+    private LocalDateTime registDate;
+    private String modifiedUserId;
```
db 상의 pk 아닌가요 ? - 그렇다면 String 이면 안될 것 같습니다.

f-lab-robert commented on src/main/java/com/flabedu/small/small/model/ItemDetail.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Getter;
+
+@Getter
+@Builder
```
Getter 와 Builder 를 사용하시는 이유가 있으실까요 ?
1. Setter 를 사용하지 않는 이유
2. Getter 를 어디서 사용하고 있는걸까요

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/ItemControllerTest.java:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    @Test
+    @DisplayName("not null 필드에 값이 전송되지 않아 상품 등록을 실패할 경우 모든 에러메세지가 응답 바디에 포함된다.")
```
이번 PR 에서는 괜찮은데요.
다음 부터는 [이 예제](https://github.com/f-lab-edu/kream/blob/development/src/test/java/com/flab/kream/order/service/OrderServiceTest.java#L141-L147)를 참고해서 조금 더 경량화 된 테스트 코드를 작성해봐주세요. 현재 작성하신 테스트 코드는 integration test 에 가깝습니다. unit test 형태의 테스트 코드를 작성하는 습관부터 들이는게 훨씬 좋습니다.

f-lab-robert commented on src/test/java/com/flabedu/small/small/controller/ItemControllerTest.java:
```diff
@@ -0,0 +1,209 @@
+                .gender(GenderEnum.C)
+                .price(new BigDecimal(20000))
+                .itemImages(Arrays.asList("img1.png","img2.png"))
+                .itemDetails(Arrays.asList(itemDetail))
```
List.of 를 사용하라고 intellij 에서 안내를 해주고 있는 것 같은데요. 
IntelliJ 의 권고를 무시하시는 이유가 있으실까요 ?

f-lab-robert commented on src/test/java/com/flabedu/small/small/service/ItemServiceTest.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    @Test
+    @DisplayName("상품 등록 시 필요한 메소드가 모두 호출된다.")
```
상품 등록시 필요한 메소드가 모두 호출되지 않거나 중간에 특정 로직이 오류가 생겼을 경우에 대한 테스트도 필요하지 않을까요 ?

f-lab-robert commented:
[ 추가 코멘트 ] 
- DTO 의 이름은 request 와 관련된건 ItemRequestDto, response 와 관련된 건 itemResponseDto 등의 형태로 작성해보시는게 어떨까요 ? 요청과 응답이 구분되는거 더 좋을 거 같습니다

KJW921222 committed:
##

KJW921222 committed:
##

KJW921222 committed:
first commit

KJW921222 committed:
##

KJW921222 committed:
##

KJW921222 committed:
##

KJW921222 committed:
## 협의 후 머지 충돌 해결

KJW921222 committed:
## 협의 후 머지 충돌 해결

KJW921222 committed:
##
